### PR TITLE
feat(#9598): add training materials page

### PR DESCRIPTION
**Pull request by latin-panda**
# Description

- Training cards in modal
  - The form title in the modal should be sticky and buttons should always be visible even when loading long content. 

- Training materials page
  - Every time a training (completed or uncompleted) is submitted, a report is recorded on the Reports list page (it's like this since before). 
  - Trainings are sorted by start date 
  - Default page is 50, scrolls to load next page (see test video)
  - Don't show the modal on top of the Training materials page 
  - The training material option is always displayed on the hamburger menu, regardless of whether a project has or does not have trainings for CHWs
  - Trainings can be opened multiple times a day (while trainings in modal display once a day) 

- Training materials page and training cards in modal
  - If the title isn't defined in the form, show the form ID
  - Show a confirmation message when quitting the training. The message and title is the same for both (page and modal) 
  - Should work for old nav and new nav display 

<details><summary>Overall functionality - happy path</summary>

https://github.com/user-attachments/assets/03220fa1-52b6-4408-bdc5-a2a7420f9eff

</details>

<details><summary>Pagination on list</summary>

https://github.com/user-attachments/assets/0b2eee9e-51d0-4fea-9bbc-ff44181f722a

</details>

#9598

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs
If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:training-page/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:training-page/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:training-page/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: api/resources/translations/messages-fr.properties [hunk 1]
```diff
@@ -1264,7 +1264,10 @@ training_cards.confirm.title = Quitter l'entraînement?
 training_cards.error.loading = Erreur lors du chargement de la formation. Veuillez contacter votre superviseur.
 training_cards.error.save = Erreur lors de l'enregistrement de la formation.
 training_cards.form.saved = Formation terminée.
-training_cards.modal.title = Changements importants
+training_materials.page.no_more_trainings = Plus de formations
+training_materials.page.no_selected = Aucun matériel de formation sélectionné
+training_materials.page.no_trainings = Aucune formation trouvée
+training_materials.page.title = Matériel de formation
```
Comment (latin-panda) [inline api/resources/translations/messages-fr.properties:1270]: @andrablaj, is this good french?

File: api/resources/translations/messages-ne.properties [hunk 2]
```diff
@@ -1264,7 +1264,10 @@ training_cards.confirm.title = तालिम छोड्ने हो?
 training_cards.error.loading = तालिम लोड गर्दा त्रुटि भयो। तपाइँको सुपरभाइजरलाई सम्पर्क गर्नुहोस्।
 training_cards.error.save = तालिम सेभ गर्न त्रुटि।
 training_cards.form.saved = तालिम सम्पन्न भयो।
-training_cards.modal.title = महत्वपुर्ण परिवर्तनहरु
+training_materials.page.no_more_trainings = अब कुनै तालिम छैन
+training_materials.page.no_selected = कुनै तालिम सामग्री छनोट गरिएको छैन।
+training_materials.page.no_trainings = कुनै तालिम फेला परेन
+training_materials.page.title = प्रशिक्षण सामग्री
```
Comment (latin-panda) [inline api/resources/translations/messages-ne.properties:1270]: @sugat009 is this good Nepali translation?

File: api/resources/translations/messages-fr.properties [hunk 3]
```diff
@@ -1264,7 +1264,10 @@ training_cards.confirm.title = Quitter l'entraînement?
 training_cards.error.loading = Erreur lors du chargement de la formation. Veuillez contacter votre superviseur.
 training_cards.error.save = Erreur lors de l'enregistrement de la formation.
 training_cards.form.saved = Formation terminée.
-training_cards.modal.title = Changements importants
+training_materials.page.no_more_trainings = Plus de formations
```
Comment (andrablaj) [inline api/resources/translations/messages-fr.properties:1267]:
```suggestion
training_materials.page.no_more_trainings = Plus aucune formation
```

@m5r does this suggestion sound OK for you?

File: api/resources/translations/messages-fr.properties (hunk 1 above)
Comment (andrablaj) [reply api/resources/translations/messages-fr.properties:1270]: I added a suggestion and tagged @m5r. All the rest sound good to me.

File: api/resources/translations/messages-fr.properties (hunk 3 above)
Comment (m5r) [reply api/resources/translations/messages-fr.properties:1267]: This is correct and removes the potential confusion between "more trainings" and "no more trainings" from the "plus" but it sounds a little bit informal. I would have said `Aucune formation restante`.
@latin-panda this literally translates to `No training left`

File: api/resources/translations/messages-fr.properties (hunk 1 above)
Comment (m5r) [reply api/resources/translations/messages-fr.properties:1270]: +1 all the rest sound good! :+1:

Comment (latin-panda) [reply api/resources/translations/messages-fr.properties:1270]: Thank you!

File: api/resources/translations/messages-fr.properties (hunk 3 above)
Comment (latin-panda) [reply api/resources/translations/messages-fr.properties]: Aucune formation restante
Okay!

File: api/resources/translations/messages-ne.properties (hunk 2 above)
Comment (sugat009) [reply api/resources/translations/messages-ne.properties:1270]: a bit of change
"no more trainings" = "थप तालिम छैन"
"no_trainings" = "तालिम फेला परेन"
"title" = "तालिम शीर्षक"

Comment (latin-panda) [reply api/resources/translations/messages-ne.properties:1270]: Thanks Sugat!

File: webapp/tests/karma/ts/modals/training-cards/training-cards.component.spec.ts
Comment (latin-panda) [inline webapp/tests/karma/ts/modals/training-cards/training-cards.component.spec.ts:1]: I'm currently working on enabling this test

**Comment by latin-panda** [comment]
@Benmuiruri, please review; this is still a work in progress. I'm still working on the test coverage.

**Review by Benmuiruri — CHANGES_REQUESTED** [review]
Nice! some neat refactoring in here. 

Just a couple of style suggestions.

**Comment by latin-panda** [comment]
@Benmuiruri @kapelo I've finished the unit and e2e test coverage. Can you please review it?

**Review by Benmuiruri — APPROVED** [review]
LGTM! 

The 3 suggestions are all optional and not a blocker.

File: webapp/src/css/modal.less [hunk 4]
```diff
@@ -101,17 +101,31 @@ mm-modal-layout {
...
+      top: 0;
+      z-index: 5;
+      background: @form-background-color;
+      width: 100%;
     }
 
     .form-footer {
-      padding: 24px 0;
+      position: absolute;
+      bottom: 0;
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/css/modal.less:125]: No, in the call, we decided to keep the scroll indicator but make the modal bigger. The indicator is helpful to know how much content is there and that it's scrollable. 
It overlaps the title and buttons due to the Enketo HTML structure, I haven't found a nicer way make the title and buttons sticky

[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/css/modal.less:123]: I found a fix, and it seems okay now
https://github.com/user-attachments/assets/1bde50af-e41b-4097-89ef-fdf0c5b1f3f2

**Comment by latin-panda** [comment]
<details><summary>Test: No trainings available</summary>

Screenshot 2024-11-14 at 6 52 08 PM
Screenshot 2024-11-14 at 6 51 57 PM

</details> 

<details><summary>Test: Error loading form</summary>

Screenshot 2024-11-14 at 6 53 20 PM

Screenshot 2024-11-14 at 6 54 48 PM

</details> 

<details><summary>Test: Loads other Enketo widgets and saves training in the reports list page.</summary>

We dont expect people to use input fields in trainings because trainings are more informative (text, images), but still good that they work fine. 


https://github.com/user-attachments/assets/d0c01a39-ee90-480e-ac14-e454fff4551c



</details>

**Comment by latin-panda** [comment]
@kapelo I have uploaded testing videos, do you have any feedback?

File: api/resources/translations/messages-en.properties [hunk 5]
```diff
@@ -1264,7 +1264,10 @@ training_cards.confirm.title = Leave training?
 training_cards.error.loading = Error loading training. Please contact your supervisor.
 training_cards.error.save = Error saving training.
 training_cards.form.saved = Training completed.
-training_cards.modal.title = Important changes
+training_materials.page.no_more_trainings = No more trainings
+training_materials.page.no_selected = No training material selected
+training_materials.page.no_trainings = No trainings found
+training_materials.page.title = Training Materials
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply api/resources/translations/messages-en.properties:1270]: Ah good eye! Thanks

File: tests/e2e/default/training-materials/training-materials.wdio-spec.js [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+      'action_icons/action_icons_note_1:label"]',
+    );
+    expect(nextCard).to.equal('The "New Action" icon at the bottom of your app has also changed.');
+    const lastCard = await trainingCardsPage.getNextCardContent(FIRST_TRAINING_NAME, 'ending/ending_note_1:label"]');
+    expect(lastCard).to.equal('If you do not understand these changes, please contact your supervisor.');
+    await trainingCardsPage.submitTraining(false);
+
+    const allTrainings = await trainingCardsPage.getAllTrainingsText();
+    expect(allTrainings.length).to.equal(2);
+    expect(allTrainings).to.have.members([ FIRST_TRAINING_NAME, SECOND_TRAINING_NAME ]);
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply tests/e2e/default/training-materials/training-materials.wdio-spec.js:107]: Thanks for the assistance! I have added this to both test cases

**Pull request merged**